src/storage/operations/mv.rs

```diff
@@ -0,0 +1,180 @@
+    operator: Operator,
+}
+
+impl OpenDalMover {
+    /// Create a new copier with the given OpenDAL operator.
+    pub fn new(operator: Operator) -> Self {
+        Self { operator }
+    }
+
+    /// Hybrid directory detection for object storage: stat first; if not available, probe prefix.
+    async fn is_directory(&self, path: &str) -> bool {
+        match self.operator.stat(path).await.ok().map(|m| m.mode()) {
+            Some(EntryMode::DIR) => true,
+            Some(_) => false,
+            None => {
+                let probe = ensure_trailing_slash(path);
+                self.operator
+                    .list_with(&probe)
+                    .limit(1)
+                    .await
+                    .map(|entries| !entries.is_empty())
+                    .unwrap_or(false)
+            }
+        }
+    }
+
+    /// Ensure a remote directory exists (appends trailing '/').
+    async fn ensure_directory(&self, dir_path: &str) -> Result<()> {
+        let to_create = ensure_trailing_slash(dir_path);
+        self.operator.create_dir(&to_create).await?;
+        Ok(())
+    }
+
+    /// Move files recursively with directory structure preservation.
+    #[async_recursion]
+    async fn move_file_recursive(&self, src_path: &str, dest_path: &str) -> Result<()> {
+        let lister = self.operator.lister_with(src_path).recursive(true).await?;
+
+        let mut stream = lister;
+        while let Some(entry) = stream.try_next().await? {
+            let meta = entry.metadata();
+            let entry_path = entry.path();
+
+            // Skip creating the root directory itself; caller ensured destination root exists
+            let src_norm = src_path.trim_start_matches('/');
+            let entry_norm = entry_path.trim_start_matches('/');
+            if meta.mode() == EntryMode::DIR && entry_norm == src_norm {
+                continue;
+            }
+
+            let relative_path = get_root_relative_path(entry_path, src_path);
+            let new_dest_path = build_remote_path(dest_path, &relative_path);
+
+            if meta.mode() == EntryMode::DIR {
+                self.ensure_directory(&new_dest_path).await?;
+            } else {
+                self.stream_move(entry_path, &new_dest_path).await?;
+                self.operator.delete(entry_path).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Stream move a single file with progress reporting.
+    async fn stream_move(&self, src_path: &str, dest_path: &str) -> opendal::Result<()> {
+        let metadata = self.operator.stat(src_path).await?;
+        let file_size = metadata.content_length();
+
+        let mut writer = self.operator.writer(dest_path).await?;
+        let mut total_bytes = 0u64;
+        let mut offset = 0u64;
+
+        let reporter = ConsoleProgressReporter::new(
+            format!("Moving {src_path}"),
+            Some(file_size),
+            DEFAULT_CHUNK_SIZE as u64,
+        );
+
+        loop {
```

Comment:
Early-exit the copy loop when offset >= file_size to avoid a 0-byte read. Not necessary after the offset fix above, but you can add:


```suggestion
        loop {
            if offset >= file_size {
                break;
            }

```